airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py

```diff
@@ -0,0 +1,463 @@
+            },
+        )
+
+        return params
+
+    def _state_filter(self, stream_state: Mapping[str, Any]) -> Mapping[str, Any]:
+        """Works differently for insights, so remove it"""
+        return {}
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        """Add fields from breakdowns to the stream schema
+        :return: A dict of the JSON schema representing this stream.
+        """
+        schema = ResourceSchemaLoader(package_name_from_class(self.__class__)).get_schema("ads_insights")
+        schema["properties"].update(self._schema_for_breakdowns())
+        return schema
+
+    @cached_property
+    def fields(self) -> List[str]:
+        """List of fields that we want to query, for now just all properties from stream's schema"""
+        schema = ResourceSchemaLoader(package_name_from_class(self.__class__)).get_schema("ads_insights")
+        return list(schema.get("properties", {}).keys())
+
+    def _schema_for_breakdowns(self) -> Mapping[str, Any]:
+        """Breakdown fields and their type"""
+        schemas = {
+            "age": {"type": ["null", "integer", "string"]},
+            "gender": {"type": ["null", "string"]},
+            "country": {"type": ["null", "string"]},
+            "dma": {"type": ["null", "string"]},
+            "region": {"type": ["null", "string"]},
+            "impression_device": {"type": ["null", "string"]},
+            "placement": {"type": ["null", "string"]},
+            "platform_position": {"type": ["null", "string"]},
+            "publisher_platform": {"type": ["null", "string"]},
+        }
+        breakdowns = self.breakdowns[:]
+        if "platform_position" in breakdowns:
+            breakdowns.append("placement")
+
+        return {breakdown: schemas[breakdown] for breakdown in self.breakdowns}
+
+    def _date_ranges(self, stream_state: Mapping[str, Any]) -> Iterator[dict]:
+        """Iterate over period between start_date/state and now
+
+        Notes: Facebook freezes insight data 28 days after it was generated, which means that all data
+            from the past 28 days may have changed since we last emitted it, so we retrieve it again.
+        """
+        state_value = stream_state.get(self.cursor_field)
+        if state_value:
+            start_date = pendulum.parse(state_value) - self.lookback_window
+        else:
+            start_date = self._start_date
+        end_date = pendulum.now()
+        start_date = max(end_date - self.INSIGHTS_RETENTION_PERIOD, start_date)
+
+        for since in pendulum.period(start_date, end_date).range("days", self._days_per_job):
+            until = min(since.add(days=self._days_per_job - 1), end_date)  # -1 because time_range is inclusive
+            yield {
+                "time_range": {"since": since.to_date_string(), "until": until.to_date_string()},
+            }
+
+    @backoff_policy
+    def _get_insights(self, params) -> AdReportRun:
```

Comment:
can we rename it to `_create_insights_job` or something that indicates it is creating a job?